server/supernova/audio_backend/portaudio_backend.hpp

```diff
@@ -124,12 +104,40 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
     bool open_stream(std::string const& input_device, unsigned int inchans, std::string const& output_device,
                      unsigned int outchans, unsigned int samplerate, unsigned int pa_blocksize, int h_blocksize) {
         int input_device_index, output_device_index;
-        if (!match_device(input_device, input_device_index) || !match_device(output_device, output_device_index))
+
+        input_device_index = GetPaDeviceFromName(input_device.c_str(), true);
+        output_device_index = GetPaDeviceFromName(output_device.c_str(), false);
+
+        // check device selection, select defaults if needed
+        std::cout << std::endl;
+        PaError checked =
+            TryGetDefaultPaDevices(&input_device_index, &output_device_index, inchans, outchans, samplerate);
+        std::cout << std::endl;
```

Comment:
remove these extra blank lines; i know it's tempting to try to make output more readable this way, but the only messages that should be set off by a blank line are error messages -- and even then, only if they're hard to distinguish from surrounding text. the reason is that this quickly becomes a slippery slope of putting a blank line between every message, or some inconsistent mix of blank lines and no blank lines that makes it look like output groupings are more meaningful then they really are, and then program output becomes unwieldy to read and navigate.

Reply:
This is now similar to what scsynth w/portaudio does. I'll write a larger comment about that below.

Also please note that, especially on Windows, device list might be quite long (we use 5 APIs...) and some visual separation might be useful.